### PR TITLE
chore: factorize condition result code out of resource plugins

### PR DESCRIPTION
**Pull request by dduportal**
This PR factorizes condition result code out of resource plugins:

- Simplify code for plugins (easier for contributors to write/update resources as they have less things to handle and can focus on the resource code)
- Centralize result object configuration (we can now change resource condition behavior in the core without impacting resources)
- Separate condition result (true or false) from error (unexpected behavior such as network error) 
- Introduces the following behavioral changes (side effects of the PR):
  -  `githubrelease` condition returns `false` but no error anymore when no upstream errors are triggered, with the same textual output as before.
  -  `gittag` condition returns `false` but no error anymore when no upstream errors are triggered, with the same textual output as before.
  -  `golang` condition returns `false` but no error anymore when no upstream errors are triggered, with the same textual output as before
  -  `gomod` condition returns `false` but no error anymore when no upstream errors are triggered, with the same textual output as before
  -  `gomodule` condition returns `false` but no error anymore when no upstream errors are triggered, with the same textual output as before
  - (nit) Homogenize methods to return an error (`error.new()` vs. `fmt.Errorf()`)
  - (cleanup) removing obsolete "ConditionFromScm" tests

## Test

To test this pull request, you can run the following commands:

```shell
make test
```

## Additional Information

- Unit test are not checking for the message returned by the `Condition()` function. It wasn't before this PR, so I chose to focus on the refactor to avoid too much changes in this PR

### CODE REVIEW

**Comment by mcwarman** [comment]
Going to continue to look more at the code as this is the sort of PR that takes multiple passes.

But I remember when I first looked to create a contribution I read this: https://github.com/updatecli/updatecli/blob/main/CONTRIBUTING.adoc#3-respect-the-contract

I realised it was out of date and moved on, in hind sight I should have looked to to update it. The whole page possibly needs a overhaul so might be out of scope for this change but contacts could be updated.

File: pkg/plugins/resources/awsami/condition.go [hunk 1]
```diff
@@ -7,21 +7,21 @@ import (
 
 	"github.com/sirupsen/logrus"
 	"github.com/updatecli/updatecli/pkg/core/pipeline/scm"
-	"github.com/updatecli/updatecli/pkg/core/result"
 )
 
 // Condition tests if an image matching the specific filters exists.
-func (a *AMI) Condition(source string, scm scm.ScmHandler, resultCondition *result.Condition) error {
+func (a *AMI) Condition(source string, scm scm.ScmHandler) (pass bool, message string, err error) {
+	// func (a *AMI) Condition(source string, scm scm.ScmHandler, resultCondition *result.Condition) error {
```
Comment (mcwarman) [inline pkg/plugins/resources/awsami/condition.go:14]: Comment left over from refactor?

Comment (dduportal) [reply pkg/plugins/resources/awsami/condition.go:14]: Good catch, thanks!

**Comment by dduportal** [comment]
> Going to continue to look more at the code as this is the sort of PR that takes multiple passes.
> 
> But I remember when I first looked to create a contribution I read this: https://github.com/updatecli/updatecli/blob/main/CONTRIBUTING.adoc#3-respect-the-contract
> 
> I realised it was out of date and moved on, in hind sight I should have looked to to update it. The whole page possibly needs a overhaul so might be out of scope for this change but contacts could be updated.

Good point! I've tried to update a bit in https://github.com/updatecli/updatecli/pull/1775/commits/0d845e01c6046bfb29d86dc64dd43b74679ced6b, WDYT?

File: pkg/plugins/resources/awsami/condition.go (hunk 1 above)
Comment (dduportal) [reply pkg/plugins/resources/awsami/condition.go]: Fixed by https://github.com/updatecli/updatecli/pull/1775/commits/f91437925c08f3ae0065eec232a5ada64b0f70d4

File: CONTRIBUTING.adoc [hunk 2]
```diff
@@ -110,7 +110,7 @@ A short one-line description of the parameter
 
 ===== 3. Respect the contract
 
-Your 'packageName' must respect at least one of the following interface contract by defining appropriated functions.
+Your 'packageName' must provides a `struct` which implements the `Resource` interface by defining the following functions:
```
Comment (olblak) [inline CONTRIBUTING.adoc:113]:
```suggestion
Your 'packageName' must provide a `struct` which implements the `Resource` interface by defining the following functions:
```

File: pkg/plugins/resources/awsami/condition_test.go [hunk 3]
```diff
@@ -21,18 +20,17 @@ func TestCondition(t *testing.T) {
 		d.ami.apiClient = mockDescribeImagesOutput{
 			Resp: d.mockedResponse,
 		}
-		gotResult := result.Condition{}
-		err := d.ami.Condition("", nil, &gotResult)
+		// gotResult := result.Condition{}
```
Comment (olblak) [inline pkg/plugins/resources/awsami/condition_test.go:23]:
```suggestion
```

Shouldn't this comment be removed?

**Comment by olblak** [comment]
I made another pass and I catch a few additional things. It's nice to see a lot of go code being cleanup

File: pkg/plugins/resources/awsami/condition_test.go (hunk 3 above)
Comment (dduportal) [reply pkg/plugins/resources/awsami/condition_test.go:23]: absolutely, looks like I forgot to remove it. Thanks!

**Pull request merged**